### PR TITLE
Chrome headless

### LINKED ISSUE #832: Support `chrome --headless`
## Change 1.

With Chrome 63 coming out we can make some big changes at Cypress.

1. We will be able to support the `debugger` protocol which will enable native events
2. We will be able to support `chrome --headless`

The reason we currently do not support `chrome --headless` is because Chrome does not support extensions in that mode. We use an extension to automate the browser and that's why its a no-go. By switching from the extension to the `debugger` protocol, we'll be able to get around this.

## Change 2.

Once we support chrome headless I propose the following changes to the CLI:

```shell
cypress run
```

Currently `cypress run` will launch Electron headlessly. This is confusing for users because they can't see what's going on and because they don't read our docs they don't understand that we're launching Electron instead of Chrome. Due to the browser version differences, they see failures in Electron that aren't happening in Chrome.

What I propose is to change this behavior so that `cypress run` by default will search for Chrome and launch it in `headed` mode. When no Chrome is installed, it'll fall back to Electron.

Currently we support a `--headed` option that only applies to Electron. This was done to prevent breaking changes.

This should now switch to being the default but then enabling the user to explicitly opt into the `--headless` mode for both Electron + Chrome.

This will be a breaking change and could come out in Cypress `2.x.x`.

### PR DESCRIPTION
**Pull request by flotwig**
- Closes #832 

### User facing changelog

* Support Chrome headless mode without the extension installed.
* Chrome can now be run with the `--headless` flag, by using `cypress run --headless --browser chrome`.
* ~~In `cypress run`, when Chrome is being used, it will be run headlessly by default. To run your tests using `xvfb`, without putting Chrome in headless mode, pass `--headed` to `cypress run`. This will make it run as it did before vX.X.X.~~
	* @Bkucera pointed out that making this the default is a breaking change, because people rely on extensions being installed during their tests, and `--headless` mode does not support extensions. We will make this default in v4.0.

### Additional details

|    | `cypress run` | `cypress run --headed` | `cypress run --headless`
| --- | --- | --- | ---
`--browser chrome` | headed | headed | headless
`--browser electron` | headless | headed | headless

- previously, the extension blocked this change according to #832
- now, the extension doesn't actually do anything:
	- the only function the extension still does is emit `change:cookie`: https://github.com/cypress-io/cypress/blob/8bcdecf2f961b07674a4b0addcbd640fc66fb18a/packages/extension/app/background.coffee#L18-L21
	- `change:cookie` is a no-op: https://github.com/cypress-io/cypress/blob/8bcdecf2f961b07674a4b0addcbd640fc66fb18a/packages/server/lib/automation/index.coffee#L93-L94 https://github.com/cypress-io/cypress/blob/8bcdecf2f961b07674a4b0addcbd640fc66fb18a/packages/server/lib/automation/cookies.coffee#L135-L149
- [x] make chrome still launch headed by default in the cli, but can run headless
- sadly, we must keep xvfb - **electron** does not support running without a display server
	- though this does put us a step closer to shipping a node.js version of Cypress without electron
- [x] only load the extension in chrome if running headed, only use it for styling + new tab page
- ~~set `CYPRESS_XVFB=1` to:~~ never mind, this is basically what passing '--headed' does
	- ~~[ ] enable `xvfb` for chrome~~ xvfb will always be on if needed, it is inherited from the CLI's DISPLAY variable
	- ~~[ ] not pass `--headless`~~
		- there is an existing method to filter chrome cli args, but this would make the process of using `xvfb` if issues arise much easier
		- is this really necessary?

### future work

-  record video in electron in headed mode

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? cypress-io/cypress-documentation#2287
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/desktop-gui/cypress/integration/project_nav_spec.js [hunk 1]
```diff
@@ -372,7 +372,7 @@ describe('Project Nav', function () {
 
     describe('browser with info', function () {
       beforeEach(function () {
-        this.info = 'The Electron browser is the version of Chrome that is bundled with Electron. Cypress uses this browser when running headlessly, so it may be useful for debugging issues that occur only in headless mode.'
+        this.info = 'foo info bar'
```
Comment (Bkucera) [inline packages/desktop-gui/cypress/integration/project_nav_spec.js]: nice

Comment (flotwig) [reply packages/desktop-gui/cypress/integration/project_nav_spec.js]: it was getting in the way of my search for headless, realized this doesn't need to be a real string

**Comment by Bkucera** [comment]
> By default, use Chrome headless mode in cypress run

this would be problematic if the user doesn't have Chrome installed - especially in CI

**Comment by flotwig** [comment]
> > By default, use Chrome headless mode in cypress run
> 
> this would be problematic if the user doesn't have Chrome installed - especially in CI

Sorry, that should read, "In `cypress run`, when Chrome is being used, run headlessly by default."

This does not change the default browser to Chrome (yet)

**Comment by Bkucera** [comment]
> > > By default, use Chrome headless mode in cypress run
> > 
> > 
> > this would be problematic if the user doesn't have Chrome installed - especially in CI
> 
> Sorry, that should read, "In `cypress run`, when Chrome is being used, run headlessly by default."
> 
> This does not change the default browser to Chrome (yet)

ah, suspected this was the case. Nice

**Comment by jennifer-shehane** [comment]
How does this affect the `isHeaded` and `isHeadless` keys, which have always been wonderfully clear 😅 

https://docs.cypress.io/api/cypress-api/browser.html#Cypress-browser-returns-browser-object

**Comment by flotwig** [comment]
> How does this affect the `isHeaded` and `isHeadless` keys, which have always been wonderfully clear sweat_smile

It appears to work as it should, I added an e2e test to demonstrate it in 4c28d89 

With `cypress run`, `Cypress.browser.isHeaded` is `false`.

With `cypress run --headed`, `Cypress.browser.isHeaded` is `true`.

`Cypress.browser.isHeadless` is just the opposite of `isHeaded`.

**Comment by flotwig** [comment]
Made all requested changes.

> can you verify that we run an e2e test in chrome that uses an extension? thx

Ya, looks like 3_plugins_spec tests that: https://github.com/cypress-io/cypress/blob/c314f056f0d961b8209fba8a83a5f8704688d1d0/packages/server/test/e2e/3_plugins_spec.coffee#L86-L94

File: packages/server/lib/modes/run.js [hunk 2]
```diff
@@ -1117,14 +1117,14 @@ module.exports = {
   },
 
   runSpecs (options = {}) {
-    const { config, browser, sys, headed, outputPath, specs, specPattern, beforeSpecRun, afterSpecRun, runUrl, parallel, group, tag } = options
+    options = _.defaults(options, {
```
Comment (brian-mann) [inline packages/server/lib/modes/run.js]: `_.defaults(...)` mutates options, so let's not use a reassignment here since that signals the wrong thing

**Pull request merged**